src/Microsoft.OData.Edm/Validation/ValidationRules.cs

```diff
@@ -1192,6 +1192,37 @@ public static class ValidationRules
                     }
                 });
 
+        // Add this to rule set as a breaking change for a future release.
+        /// <summary>
+        /// Validates that if a complex-typed property has a type or base type that is the same as
+        /// the complex type, then the complex type is nullable.
+        /// </summary>
+        public static readonly ValidationRule<IEdmStructuralProperty> RecursiveComplexTypedPropertyMustBeOptional =
+            new ValidationRule<IEdmStructuralProperty>(
+                (context, property) =>
+                {
+                    IEdmTypeReference currentPropTypeRef = property.Type;
+                    IEdmType currentPropType = currentPropTypeRef.Definition;
+                    IEdmStructuredType declaringType = property.DeclaringType;
+
+                    if (!currentPropTypeRef.IsNullable)
+                    {
+                        while (currentPropType != null && currentPropType != declaringType)
+                        {
+                            IEdmStructuredType baseType = ((IEdmStructuredType)currentPropType).BaseType;
+                            currentPropType = baseType;
+                        }
+
+                        if (currentPropType == declaringType)

```

Comment:
Shouldn't the error be inside the `while` loop? Something like:
```
for (; currentPropType != null; currentPropType = ((IEdmStructuredType)currentPropType).BaseType)
{
  if (currentPropType == declaringType)
  {
    context.AddError(...);
  }
}
```
This way, the caller gets two errors for something like this:
```
<ComplexType Name="foo">
  <Property Name="prop1" Type="foo" />
  <Property Name="prop2" Type="foo" />
</ComplexType>
```

Because both `prop1` nor `prop2` should be nullable

Reply:
That's a good point! I'll add a test to capture this case.

Reply:
Ok, I think I see why the `if` statement is outside of the `while` loop. Can you please add a test case for the "two property" case anyway?

Reply:
Also, take this model:
```
<ComplexType Name="foo">
  <Property Name="prop1" Type="bar" />
</ComplexType>
<ComplexType Name="bar">
  <Property Name="prop2" Type="fizz" />
  <Property Name="prop3" Type="foo" />
</ComplexType>
<ComplexType Name="fizz">
  <Property Name="prop4" Type="foo" />
</ComplexType>
```
Can you add a test to confirm that there is an error for both `prop3` and `prop4`?

Reply:
Actually, I believe this case falls under the umbrella of the general case. Because the validation rule is for a given property, rather than a given complex type, it will evaluate each of those properties individually. The `currentPropType` is cycling through the inheritance tree of types for that given property (rather than each of the properties in a given complex type).

Reply:
If I'm reading the issue correctly, it sounds like we want to make sure we capture all of these infinite recursions. Are you saying there's already validation for some of them? And are there associated tests for that validation (that match my suggestions above)?

Reply:
Added test for multiple invalid properties for completeness ✅